Command mode support (#7681)
quarkusio/quarkus: docs/src/main/asciidoc/lifecycle.adoc

* an `org.acme.lifecycle.GreetingResource` resource
* an associated test

== The main method

By default Quarkus will automatically generate a main method, that will boostrap Quarkus and then just wait for
shutdown to be initiated. Let's provide our own main method:
[source,java]
----
package om.acme;

> import io.quarkus.runtime.annotations.DefaultMain;

Reviewer: ```suggestion
import io.quarkus.runtime.annotations.QuarkusMain;
```

import io.quarkus.runtime.Quarkus;

> @DefaultMain  <1>

Reviewer: ```suggestion
@QuarkusMain  <1>
```

public class Main {

    public static void main(String ... args) {
        System.out.println("Running main method");
        Quarkus.run(args); <2>
    }
}
----
<1> This annotation tells Quarkus to use this as the main method, unless it is overridden in the config
<2> This launches Quarkus

This main class will bootstrap Quarkus and run it until it stops. This is no different to the automatically
generated main class, but has the advantage that you can just launch it directly from the IDE without needing
to run a Maven or Gradle command.

> WARNING: It is not recommenced to do any business logic in this main method, as Quarkus has not been setup yet,

Reviewer: ```suggestion
WARNING: It is not recommenced to do any business logic in this main method, as Quarkus has not been set up yet,
```

and Quarkus may run in a different ClassLoader. If you want to perform logic on startup use an `io.quarkus.runtime.QuarkusApplication`
as described below.

If we want to actually perform business logic on startup (or write applications that complete a task and then exit)
we need to supply a `io.quarkus.runtime.QuarkusApplication` class to the run method. After Quarkus has been started
the `run` method of the application will be invoked. When this method returns the Quarkus application will exit.

If you want to perform logic on startup you should call `Quarkus.waitForExit()`, this method will wait until a shutdown
is requested (either from an external signal like when you press `Ctrl+C` or because a thread has called `Quarkus.asyncExit()`).

An example of what this looks like is below:

[source,java]
----
package com.acme;

import io.quarkus.runtime.Quarkus;
import io.quarkus.runtime.QuarkusApplication;

> import io.quarkus.runtime.annotations.DefaultMain;

Reviewer: ```suggestion
import io.quarkus.runtime.annotations.QuarkusMain;
```

> @DefaultMain

Reviewer: ```suggestion
@QuarkusMain
```
